viewer/packages/tools/src/Measurement/MeasurementTool.ts

```diff
@@ -0,0 +1,444 @@
+/*!
+ * Copyright 2022 Cognite AS
+ */
+
+import { Cognite3DViewer } from '@reveal/api';
+import { Cognite3DViewerToolBase } from '../Cognite3DViewerToolBase';
+import { assertNever, DisposedDelegate, EventTrigger } from '@reveal/utilities';
+import * as THREE from 'three';
+import {
+  MeasurementAddedDelegate,
+  MeasurementStartedDelegate,
+  MeasurementEndedDelegate,
+  MeasurementOptions
+} from './types';
+import { MeasurementManager, Measurement } from './MeasurementManager';
+import { MeasurementLabels } from './MeasurementLabels';
+import { HtmlOverlayTool, HtmlOverlayToolOptions } from '../HtmlOverlay/HtmlOverlayTool';
+import rulerSvg from './styles/ruler.svg';
+
+type MeasurementEvents = 'added' | 'started' | 'ended' | 'disposed';
+
+enum MeasurementStates {
+  Default,
+  Started,
+  Added,
+  Ended
+}
```

Comment:
This seems like unnecessary and will just make things harder. Why do you need this? As far as I can see, the only needed change to fix `exitMeasurement()` on `added` was to change the order in which measurements were added to the list of measurements, and when the event was fired.

What is the difference in behaviour when in `Default` state and `Ended`? What is the behaviour in `Added` mode? I bet I can provoke bugs here - e.g. call `enterMeasurementMode()` after a measurement just has been added.